pkg/controller/baremetalhost/baremetalhost_controller.go

```diff
@@ -900,3 +950,158 @@ func (r *ReconcileBareMetalHost) publishEvent(request reconcile.Request, event c
+	if &raidConfig == nil {
+		return nil, nil
+	}
+	logicalDisks := []map[string]interface{}{}
+	if raidConfig.RootVolume != nil {
+		err = r.validateRAIDVolumeConfig(raidConfig.RootVolume)
+		if err != nil {
+			return nil, err
+		}
+		logicalDisks = append(logicalDisks, r.buildRAIDVolumeArgs(raidConfig.RootVolume, true))
+	}
+	for _, volume := range raidConfig.Volumes {
+		err = r.validateRAIDVolumeConfig(&volume)
+		if err != nil {
+			return nil, err
+		}
+		logicalDisks = append(logicalDisks, r.buildRAIDVolumeArgs(&volume, false))
+	}
+	if len(logicalDisks) > 0 {
+		cleanStep = &nodes.CleanStep{
+			Interface: "raid",
+			Step: "create_configuration",
+			Args: map[string]interface{}{"logical_disks": logicalDisks},
+		}
+	}
+	return cleanStep, nil
+}
+
+// A private method to validate incoming RAID configuration.
+func (r *ReconcileBareMetalHost) validateRAIDVolumeConfig(volumeConfig *metal3v1alpha1.RAIDVolume) (err error) {
+	// The config without RAIDLevel will be marked as invalid
+	if volumeConfig.RAIDLevel == "" {
+		return errors.New("Missing the RAID level of disk 'raidLevel' in RAID config")
+	}
+	return nil
+}
+
+// A private method to build the Args in CleanStep for RAID configuration
+func (r *ReconcileBareMetalHost) buildRAIDVolumeArgs(
+	volumeConfig *metal3v1alpha1.RAIDVolume, isRootVolume bool) map[string]interface{} {
+	volume := map[string]interface{}{}
+	if isRootVolume {
+		volume["is_root_volume"] = isRootVolume
+	}
+	if volumeConfig.SizeGB != nil {
+		volume["size_gb"] = volumeConfig.SizeGB
+	} else {
+		volume["size_gb"] = "MAX"
+	}
+	if volumeConfig.RAIDLevel != "" {
+		volume["raid_level"] = volumeConfig.RAIDLevel
+	}
+	if volumeConfig.VolumeName != "" {
+		volume["volume_name"] = volumeConfig.VolumeName
+	}
+	if volumeConfig.SharePhysicalDisks != nil {
+		volume["share_physical_disks"] = volumeConfig.SharePhysicalDisks
+	}
+	if volumeConfig.DiskType != "" {
+		volume["disk_type"] = volumeConfig.DiskType
+	}
+	if volumeConfig.InterfaceType != "" {
+		volume["interface_type"] = volumeConfig.InterfaceType
+	}
+	if volumeConfig.NumberOfPhysicalDisks > 0 {
+		volume["number_of_physical_disks"] = volumeConfig.NumberOfPhysicalDisks
+	}
+	if volumeConfig.Controller != "" {
+		volume["controller"] = volumeConfig.Controller
+	}
+	if volumeConfig.PhysicalDisks != nil {
+		volume["physical_disks"] = volumeConfig.PhysicalDisks
+	}
+	return volume
+}
+
+// ValidateAndBuildBIOSCleanSteps verify the given BIOS configuration
+func (r *ReconcileBareMetalHost) ValidateAndBuildBIOSCleanStep(
+	prov provisioner.Provisioner, biosConfig metal3v1alpha1.BIOSConfig) (cleanStep *nodes.CleanStep, err error){
+	if biosConfig == nil {
+		return nil, nil
+	}
+	accessDetails := prov.GetAccessDetails()
+	biosConfigDetails := accessDetails.GetBIOSConfigDetails()
+	if biosConfigDetails == nil {
+		return nil, errors.New(fmt.Sprintf(
+			"'%s' driver does not support BIOS configuration", accessDetails.Driver()))
```

Comment:
This is incorrect. The driver might support BIOS, we (BMO) doesn't support it.

Reply:
Thanks. I will update it :)